### ISSUE TITLE
Lazy task API migration confusion WRT lost dependency information

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#10683]
## Problem

Tim raised this issue https://github.com/gradle/gradle/issues/5938 while converting some tasks to use the new task API. 

He was replacing something like:
```
tasks.create("consumer", Tar) {
    it.from({ tasks.generate })
}

afterEvaluate {
    tasks.create('generate') {
        it.outputs.file("$buildDir/text/sentence.txt")

        it.doLast {
            file("$buildDir/text/sentence.txt").text = "Here is some text"
        }
    }
}
```

with
```
tasks.register("consumer", Tar) {
    it.from(provider { tasks.generate })
}

afterEvaluate {
    tasks.register('generate') {
        it.outputs.file("$buildDir/text/sentence.txt")

        it.doLast {
            file("$buildDir/text/sentence.txt").text = "Here is some text"
        }
    }
}
```

This fails because the dependency between consumer and generate is lost by the `provider {}`. 

There are several better alternatives:

### Use `named(...)` directly:
```
tasks.register("consumer", Tar) {
    it.from(tasks.named("generate"))
}
```

This just works and probably what we would suggest to always do.

### Use `named(...)` directly, but keep the indirection (naïve translation):
```
tasks.register("consumer", Tar) {
    it.from({ tasks.named("generate") })
}
```

This works, but it's a weird and unnecessary indirection.

### Don't change anything
```
tasks.register("consumer", Tar) {
    it.from({ tasks.generate })
}
```

This still works. The Callable was already lazy.

## Bottom line

This is something we should explain in documentation, but I think we should also consider a couple of other things:

- `provider { buildable-thing }` losing dependency information is weird. A closure/callable doesn't do that.
- Why is `provider { tasks.generate }` considered better than `{ tasks.generate }` or why is `named(...)` hard to find?

**Comment by big-guy** [comment]
This was fixed
